src/norm_flows.jl

```diff
@@ -0,0 +1,98 @@
+using Distributions
+using LinearAlgebra
+using Random
+using StatsFuns: softplus
+using Roots, LinearAlgebra # for inverse
+
+################################################################################
+#                            Planar and Radial Flows                           #
+#             Ref: Variational Inference with Normalizing Flows,               #
+#               D. Rezende, S. Mohamed(2015) arXiv:1505.05770                  #
+################################################################################
+
+mutable struct PlanarLayer <: Bijector
+    w
+    u
+    u_hat
+    b
+end
+
+mutable struct RadialLayer <: Bijector
+    α_
+    β
+    z_not
+end
+
+function update_u_hat(u, w)
+    # to preserve invertibility
+    u_hat = u + (planar_flow_m(transpose(w) * u) - transpose(w) * u)[1] * w/(norm(w[:,1],2) ^ 2)
+end
+
+function update_u_hat!(flow::PlanarLayer)
+    flow.u_hat = flow.u + (planar_flow_m(transpose(flow.w) * flow.u) - transpose(flow.w) * flow.u)[1] * flow.w/(norm(flow.w,2)^2)
+end
+
+
+function PlanarLayer(dims::Int)
+    w = param(randn(dims, 1))
+    u = param(randn(dims, 1))
+    b = param(randn(1))
+    u_hat = update_u_hat(u, w)
+    return PlanarLayer(w, u, u_hat, b)
+end
+
+function RadialLayer(dims::Int)
+    α_ = param(randn(1))
+    β = param(randn(1))
+    z_not = param(randn(dims, 1))
+    return RadialLayer(α_, β, z_not)
+end
+
+planar_flow_m(x) = -1 .+ log.(1 .+ exp.(x)) # For planar flow
+dtanh(x) = 1 .- (tanh.(x)).^2 # For planar flow
+ψ(z, w, b) = dtanh(transpose(w) * z .+ b) .* w # For planar flow
+h(α, r) = 1 ./ (α .+ r) # For radial flow
+dh(α, r) = -h(α, r).^2 # For radial flow
```

Comment:
Do we really need these shorthand functions? I feel it makes it hard to read in the main function when we see things like `h` or `dh`. Also here, at least we can replace `h` with `inv(a + r)`?

Reply:
I actually kind of like it, as it makes it very easy to match with the paper:) If someone wants to read and understand the code, close to 1-to-1 matching with paper is helpful. But this is also very subjective, so I'm not very invested in either option.

But I think for readability it might be nice to structure the entire file slightly differently. Instead of intertwining the definitions of the different methods, e.g.
```julia
transform(::PlanarFlow, z)
transform(::RadialFlow, z)

forward(::PlanarFlow, z)
forward(::RadialFlow, z)

# rest of RadialFlow and PlanarFlow impls...
```
we could do
```julia
transform(::PlanarFlow, z)
forward(::PlanarFlow, z)
# rest of PlanarFlow impls...

transform(::RadialFlow, z)
forward(::RadialFlow, z)
# rest of RadialFlow impls...
```
and so on for all the other definitions?

Reply:
@torfjelde I agree on the restructuring. Makes it easier to follow.
@xukai92 As tor said, I too feel that 1-1 correspondence with the paper will be much easier to follow. But I am not how important it is to maintain this 1-1 correspondence. I am conflicted on this.